### DISCUSSION TITLE
permissions error when building the docker image over openshift

### DISCUSSION THREAD

**Discussion by samisalamiws** [containerbase/base#312] [Q&A]
Hi @viceice 

we are having some permissions errors when we're building our dockerfile based buildpack over openshift. A lot of simple commands like creating folders are not able to run. 

we followed the workaround suggested on their documentation at: https://developers.redhat.com/blog/2020/10/26/adapting-docker-and-kubernetes-containers-to-run-on-red-hat-openshift-container-platform#group_ownership_and_file_permission but this made our image size increase significantly (over 10 GB).
basically we added these commands:
`RUN chown -R ${USER_ID}:0 /usr/local/buildpack/tools/`
`RUN chown -R ${USER_ID}:0 /usr/local/bin/`
`RUN chown -R ${USER_ID}:0 /home/wss-scanner/`
`RUN chown -R ${USER_ID}:0 /tmp`

This brings us to an older discussion we had regarding user/group creation.
when we ran the old dockerfile (without buildpack), we creating the group/user as follow:

`RUN groupadd ${WSS_GROUP} && \ `
`useradd --gid ${WSS_GROUP} --groups 0 --shell /bin/bash --home-dir ${WSS_USER_HOME} --create-home ${WSS_USER} && \passwd -d ${WSS_USER} `

when we ran the old dockerfile over openshift, it worked ok. 
but currently, we are not able to create the group/user the way we were, and we're facing permissions issues.

Is there a way to adjust the user --groups value?

**Comment by viceice** [comment]
🤔 the Buildpack user already is a member of the root group.

So can you provide some more details about which folders of buildpack you need write access to and why?

@rarkins is that the scanner image i converted some time ago? If yes, we can exchange the details in our private VisualOn repo.

**Reply by rarkins** [reply]
Yes, it's the same. Can you add @samisalamiws to the repo and we continue discussion there?

@samisalamiws can you update that repo with a PR showing any changes you've made?

**Reply by samisalamiws** [reply]
basically, every place that we're using the `RUN install-tool <package_manage>` command. for example, most of them are being installed to the `/home/wss-scanner` folder, so this folder should give the user write permissions. 
please note, there are package managers like go that are installed to another folder than the `/home/wss-scanner` (to `/go`)

from the help of the `adduser` command:
```
-g, --gid GROUP

The group name or number of the user's initial login group. The group name must exist. A group number must refer to an already existing group.

-G, --groups GROUP1[,GROUP2,...[,GROUPN]]]

A list of supplementary groups which the user is also a member of. Each group is separated from the next by a comma, with no intervening whitespace. The groups are subject to the same restrictions as the group given with the -g option. The default is for the user to belong only to the initial group.
```
maybe we need to really convert the `--groups` and `-gid` values, to what we had before? 
currently in buildpack is:
```
groupadd --gid ${USER_ID} ${USER_NAME};
useradd --uid ${USER_ID} --gid 0 --groups ${USER_NAME} --shell /bin/bash --create-home ${USER_NAME}
```

**Reply by viceice** [reply]
Invited

**Reply by viceice** [reply]
@samisalamiws We recently fixed some bugs for openshift, can you check the latest patch releases?

**Reply by samisalamiws** [reply]
@viceice I see releases with version changes. Can you be more specific? we do all the commands that are relevant to the scanner dockerfile

**Reply by rarkins** [reply]
I think he means update to latest buildpack version

**Reply by viceice** [reply]
> I think he means update to latest buildpack version 

Yes, sorry for not specific enough.

**Reply by samisalamiws** [reply]
@rarkins thanks for the update
@viceice I'm using now the latest release available `3.7.2` and the problem still happens

**Reply by rarkins** [reply]
Please provide an example Dockerfile

**Reply by samisalamiws** [reply]
@rarkins 
So currently we're facing two problems related to permissions.
1. regarding open shift
2. regarding cleaning go packages 

for both of these problems, use the following dockerfile:
```
# This default may be swapped for any compatible base image
ARG BASE_IMAGE=amd64/ubuntu:18.04

# This buildpack is used for tool installation and user/directory setup
FROM containerbase/buildpack:3.2.4@sha256:31e3b85387934e330d9fcddf793ae07c60d23e6ad3d01a070b4060594eedb8f7 AS buildpack

FROM ${BASE_IMAGE} as base


# The buildpack supports custom user
ARG APT_HTTP_PROXY

# The buildpack supports custom user
ARG USER_NAME=wss-scanner
ARG USER_ID=1000
ARG USER_HOME=/home/${USER_NAME}
ARG APT_HTTP_PROXY

# Set env and shell
ENV BASH_ENV=/usr/local/etc/env
SHELL ["/bin/bash" , "-c"]

# Set up buildpack
COPY --from=buildpack /usr/local/bin/ /usr/local/bin/
COPY --from=buildpack /usr/local/buildpack/ /usr/local/buildpack/
RUN install-buildpack

# renovate: datasource=github-tags depName=git lookupName=git/git
ARG GIT_VERSION=2.34.1
RUN install-tool git

# renovate: datasource=docker depName=golang versioning=docker
ARG GOLANG_VERSION=1.17.6
RUN install-tool golang

USER ${USER_ID}
## Install package managers
RUN go install github.com/tools/godep@latest
RUN go install github.com/LK4D4/vndr@latest
RUN go install  github.com/kardianos/govendor@latest

### Switch User ###
ENV HOME ${USER_HOME}
WORKDIR ${USER_HOME}
USER ${USER_ID}
```

for the go issue, run the following commands after installing the image:
```
touch $HOME/timestamp
go get github.com/gookit/color
find /go -newer $HOME/timestamp -type f -exec rm -rf {} +
```

for openshift, simply do:
`mkdir /go/pkg/mod/cache/download/github.com/smallnest`
in openshift it's happening for more package managers, like maven as well

**Reply by samisalamiws** [reply]
We tried also `RUN usermod -a -G 0 ${USER_NAME} ` but it didn't help
The only thing that helped for the openshift, was by adding these commands:
```
RUN chown -R ${USER_ID}:0 /home/wss-scanner/
RUN chown -R ${USER_ID}:0 /go
```
but as I mentioned before, it made the image to double it size, plus it didn't helped with the `find` command mentioned in my previous comment

**Reply by viceice** [reply]
Ok, it seems some folders in home are getting wrong permissions when you run commands as user 1000 and then with an arbitrary user id on openshift.

Can you create two minimal docker files which will fail. For openshift it should be enough to change to a random user id and run the failing commands.

With those reproduction we should be able to identify the wrong folders.

**Reply by rarkins** [reply]
If we add a recursive chown of home directory to `1000:0` at end the of the Dockerfile, does it fix it?

**Reply by viceice** [reply]
Yes, but that would copy all files to a new docker layer and duplicate size. See comments of @samisalamiws above

**Reply by samisalamiws** [reply]
@rarkins for openshift yes, but then the image will be 20 GB instead of ~10 GB now

@viceice the dockerfile that I pasted on my previous comment does reproduce the problem. Do you want me to create another one?

**Reply by viceice** [reply]
I assume that the go stuff is creating folders with the permissions mentioned.

**Reply by rarkins** [reply]
Can our `install-tool` wrapper do a home directory `chown` automatically? That way it's not *adding* any layers

**Reply by viceice** [reply]
> @viceice the dockerfile that I pasted on my previous comment does reproduce the problem. Do you want me to create another one?

Ok, so we can try to add that to our tests here and try to fix the issue.

**Reply by viceice** [reply]
> Can our `install-tool` wrapper do a home directory `chown` automatically? That way it's not *adding* any layers

It already does that. But if any external tool call, like `go install`, is creating new folders in home, they will have wrong permissions for openshift.

**Reply by rarkins** [reply]
Do you think you can change the behavior of `go`? Or should @samisalamiws add a `chown` in the same RUN command as he runs `go install`?

**Reply by viceice** [reply]
I think the current best solution is to use a single RUN for the `go install` and do the chown on newly created folders only.

I don't think I can easily change go behavior, but maybe the `sgid` bit can force all future subfolders to inherit group. @Chumper can you elaborate this?

**Reply by samisalamiws** [reply]
@viceice but the problem is not just for go (in openshift). 
what do you mean by "single RUN"?

**Reply by viceice** [reply]
A single docker `RUN` command in docker file.

**Reply by samisalamiws** [reply]
@viceice @rarkins We're thinking of doing a user change (to our user that we created in the Dockerfile and gave the needed permissions), before each scan that we trigger. so the `su {USER_NAME}` command will be run from our code.
The problem here is that we don't have the password to do so. Is it possible to give this user a password that will be saved as a config param? What do you think about this approach?

**Reply by viceice** [reply]
You would need to set the password manually, as we don't want to have any passwords safer to buildpack images.

We still should identify the tools which are causing issues with openshift.

Also your use case is different to our other known cases. As you add stuff as buildpack user and then try to do the same as arbitrary openshift user. Until now we only had one of them.

**Reply by rarkins** [reply]
Would it help if we add a util command "chown-home" that can be appended to RUN commands? The challenge though is to support this stuff at runtime too.

BTW we could try to coerce go into installing into home - that might help

**Reply by viceice** [reply]
> Would it help if we add a util command "chown-home" that can be appended to RUN commands? The challenge though is to support this stuff at runtime too.

This needs to be done selective, as otherwise we duplicate files again.


> BTW we could try to coerce go into installing into home - that might help

https://github.com/containerbase/buildpack/blob/ac6a459922bcc1d2f23b499bad615a399d30d383/src/usr/local/buildpack/tools/golang.sh#L34-L44

**Reply by samisalamiws** [reply]
@rarkins yes, putting the go into the home directory will be very helpful, plus making sure that any new tool (package manager) installation will be done also to the home directory

@viceice How can we change a password for a user that has already been created? because currently, the buildpack is the one creating the user.
the tools that are causing the problem is all the tools that try to create folders/files in the home directory, for example:
maven - when we do `mvn build install` it downloads dependencies to the cache directory which is located in the ${HOME}/.m2 folder
go - the same, except that the folder is located in `/go` 
npm - when installing global packages (note it's not happening for local packages, as they are installed to a directory in the `/tmp` which the current arbitrary user creates)
I don't remember all of them, but I can do some research about it

by the way, after cleaning some of the unnecessary commands, we got a docker image of size 14.5 GB instead of 20GB. so we might live with it for the next couple of weeks until we have a more profound solution.

**Reply by viceice** [reply]
Can you set `UMASK=770` when running user commands on docker build? This should make all new files and folders writable by primary group.

**Comment by viceice** [comment]
@samisalamiws I've fixed a number of wrong permissions on home dir. Can you check wich folder are still wrong?

Please add observations to:
- #517
